### PR TITLE
S3655: Implement the rule

### LINKED ISSUE #6682: Fix S3655 FP: HasValue checked with Pattern matching
Reported here: https://community.sonarsource.com/t/rule-c-s3655-false-positive-for-pattern/81703

```
public struct Repro_6682
{
    public bool SomeProperty { get; }

    public void PatternMatching(Repro_6682? arg, bool condition)
    {
        if (condition)
        {
            arg = null;
        }

        if (arg is { SomeProperty: true })   // A null check
        {
            var value = arg.Value;  // Noncompliant FP
        }

        if (arg is { })   // A null check
        {
            var value = arg.Value;  // Noncompliant FP
        }
    }
}
```

### PR DESCRIPTION
**Pull request by antonioaversa**
Task 4 of https://github.com/SonarSource/sonar-dotnet/issues/6794

Previous task: https://github.com/SonarSource/sonar-dotnet/pull/6909

Fixes #6682

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 1]
```diff
@@ -2,132 +2,110 @@
...
-        if (i1.HasValue)
-        {
-            Console.WriteLine(i1.Value);
-        }
-
-        Console.WriteLine(i1.Value); // FIXME Non-compliant {{'i1' is null on at least one execution path.}}
-        // FIXME          ~~~~~~~~
-    }
+    IEnumerable<TestClass> EnumerableExpressionWithCompilableCode() =>
+        Numbers.OrderBy(x => x.Number.HasValue).ThenBy(i => i.Number);
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:28]: This is out of scope right? If so, the test case should be removed or at least an appropriate comment to be added.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 2]
```diff
@@ -2,132 +2,110 @@
...
 
-    public IEnumerable<TestClass> TestEnumerableExpressionWithCompilableCode() => numbers.OrderBy(i => i.Number.HasValue).ThenBy(i => i.Number);
-    public IEnumerable<int> TestEnumerableExpressionWithNonCompilableCode() => numbers.OrderBy(i => i.Number.HasValue).ThenBy(i => i.Number).Select(x => x.Number ?? 0);
+    IEnumerable<int> EnumerableExpressionWithNonCompilableCode() =>
+        Numbers.OrderBy(x => x.Number.HasValue).ThenBy(i => i.Number).Select(x => x.Number ?? 0);
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:31]: Same as above. But it also confuses me: This compiles for me, so why the name?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 3]
```diff
@@ -168,133 +150,254 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
+                }
+                if (!foo.HasValue)
+                {
+                    Console.WriteLine(foo.Value.ToString());
+                }
+                if (foo == null)
+                {
+                    // FIXME the following
+                    Console.WriteLine(foo.Value.ToString());    // FN
+                }
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:376]: Why is this a FN? Looks compliant to me, since the code is unreachable.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 4]
```diff
@@ -168,133 +150,254 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
+                    Console.WriteLine(foo.Value.ToString());    // Compliant, unreachable
+                }
+                if (!foo.HasValue)
+                {
+                    Console.WriteLine(foo.Value.ToString());    // FIXME Non-compliant
+                }
+                if (foo == null)
+                {
+                    Console.WriteLine(foo.Value.ToString());    // FN
+                }
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:395]: These are basically the same cases. Why do they not have the same comments?

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
The implementation doesn't look super elegant, but I'm unaware of a better solution.
Besides that, there is Just a bit of confusion in the tests.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 1 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:28]: Indeed, I couldn't find how the test could relate to S3655.

I have changed it to make it more relevant, and moved to a class `Linq`, which is supposed to test nullable value types use in Linq expressions.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 2 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:31]: I don't know, it was there in the previous testing code (for the Sonar version of the symbolic execution), and I hadn't removed it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 4 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:395]: Indeed. I have aligned them, keeping into account that `value.HasValue` make them somewhat antisimmetric:
- compliance due to non-emptiness becomes compliance due to unreachability
- compliance due to unreachability becomes non-compliance due to emptiness

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 5]
```diff
@@ -270,6 +251,28 @@ class TernaryOperator
...
+        new TestClass(),
+        new TestClass() { Number = 1 },
+        new TestClass() { Number = null }
+    };
+
+    IEnumerable<int> EnumerableOfEmptyNullableValues1 =>
+        Numbers.Where(x => !x.Number.HasValue).Select(x => x.Number.Value); // FN
+
+    IEnumerable<int> EnumerableOfEmptyNullableValues2 =>
+        Numbers.Select(x => null as int?).Select(x => x.Value);             // FN
```
Comment (Tim-Pohlmann) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:268]: NIT: I wouldn't even consider these FN. For me they are just out of scope.

**Review by Tim-Pohlmann — APPROVED** [review]
LGTM

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Rebase is needed after rebase of feature branch

**Comment by antonioaversa** [comment]
> Rebase is needed after rebase of feature branch

@pavel-mikula-sonarsource The branch is on top of `feature/SE`. 
The rebase was done while you were writing the comment :-)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.CSharp9.cs [hunk 6]
```diff
@@ -66,16 +66,25 @@ public void PatternMatching(int? arg)
             {
                 v = nullable.Value;
             }
-            v = nullable.Value;        // FN, can't build CFG for this method
+            v = nullable.Value;        // Noncompliant
         }
     }
 
+    void PatternMatching(int? i, double? d, float? f)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.CSharp9.cs:73]: Unused
```suggestion
    void PatternMatching(int? i)
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.CSharp9.cs [hunk 7]
```diff
@@ -66,16 +66,25 @@ public void PatternMatching(int? arg)
...
+    {
+        if (i.HasValue is true && i.Value == 42) { }
+        if (i is not null && i.Value == 42) { }
+
+        if (i.HasValue is not true && i.Value == 42) { } // Noncompliant
+        if (i is null && i.Value == 42) { } // Noncompliant
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.CSharp9.cs:79]: This would read easier when aligned to the same tab distance
```suggestion
        if (i.HasValue is not true && i.Value == 42) { } // Noncompliant
        if (i is null && i.Value == 42) { }              // Noncompliant
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.CSharp9.cs [hunk 8]
```diff
@@ -140,12 +149,12 @@ public void PatternMatching(Repro_6682? arg, bool condition)
 
         if (arg is { SomeProperty: true })  // A null check
         {
-            var value = arg.Value;  // FIXME Non-compliant FP
+            var value = arg.Value;          // Compliant
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.CSharp9.cs:152]: As this explicitly fixes a reported FP, please update the description to contain `Fixes #6682`. The automation will start moving the issue around with next actions. To catch up manually:
* drag it in "In progress" column, it will put your face on it
* remove this PR from Kanban, as the issue will represent it
* remove assignee from the PR, as the issue will represent it

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 9]
```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
-public void NestedCombined(DateTime? value)
-{
-    if (foo == value)   // Relationship is added here
+    public void Sequence(DateTime? value)
     {
         if (value.HasValue)
+        {
+            // HasValue and NoValue constraints are set here
+        }
+        if (foo == value)   // Relationship is added here
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:325]: As relationships are not supported by the engine, we need to update all these comments (also below)
```suggestion
        if (foo == value)   // Relationship should be created here
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 10]
```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
 
-public interface IWithDefaultImplementation
+class NullableOfCustomTypes
 {
-    int DoSomething()
+    void Assignment(AStruct? nullable)
     {
-        int? i = null;
-        return i.Value; //FIXME Non-compliant
+        _ = nullable.Value;                                              // Compliant, unknown
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:151]: Aesthetics: It would be nicer to indent the comments something like -20 spaces in this method.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 11]
```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
+        nullable = new Nullable<AStruct>();
+        _ = nullable.Value;                                              // Noncompliant
+
+        _ = (new Nullable<AStruct>()).Value;                             // Noncompliant
+        _ = (null as Nullable<AStruct>).Value;                           // Noncompliant
+
+        nullable = null;
+        _ = ((AStruct?)nullable).Value;                                  // Noncompliant
+        _ = (nullable as AStruct?).Value;                                // FN
+        _ = ((AStruct?)(nullable as AStruct?)).Value;                    // FN
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:173]: Please re-assign `nullable=null` before each line to make sure we didn't learn the NotNull from using `.Value`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 12]
```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
+    bool BitAndLearningNonNull(bool? b) => b.HasValue & b.Value;            // Noncompliant - i.Value reached in any case
+
+    bool LogicalOrLearningNonNull(bool? b) => b.HasValue || b.Value;        // Noncompliant - i.Value reached when i.HasValue is false
+
+    bool BitOrLearningNonNull(bool? b) => b.HasValue | b.Value;             // Noncompliant - i.Value reached in any case
+
+    bool Tautology(bool? b) => (!b.HasValue || b.HasValue) && b.Value;      // Noncompliant
+
+    bool ShortCircuitedOr(bool? b) => (true || b.HasValue) && b.Value;      // Compliant
+    bool ShortCircuitedAnd(bool? b) => (false &&  b.HasValue) || b.Value;   // Compliant
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:231]: These might be 'FN` that we don't want to fix. As user indicated it could be null, and accesses it in any case.

S2583 should kick in to warn users about the problem itself.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 13]
```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
+{
+    bool IndependentConditions1(double? d, float? f) => f == null && d.Value == 42.0;  // Compliant, f imposes no constraints on d
+    bool IndependentConditions2(double? d, float? f) => f != null && d.Value == 42.0;  // Compliant
+    bool IndependentConditions3(double? d, float? f) => f.HasValue && d.Value == 42.0; // Compliant
+    bool IndependentConditions4(double? d, float? f) => null == f && d.Value == 42.0;  // Compliant
+    bool IndependentConditions5(double? d, float? f) => null == f && d.Value == 42.0;  // Compliant
+
+    bool DependentConditions1(double? d, float? f) =>
+        !d.HasValue && d.Value == 42.0;                                 // Noncompliant
+    bool DependentConditions2(double? d, float? f) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:247]: Unused
```suggestion
    bool DependentConditions1(double? d) =>
        !d.HasValue && d.Value == 42.0;                                 // Noncompliant
    bool DependentConditions2(double? d) =>
```

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 14]
```diff
@@ -36,6 +38,33 @@ public override bool ShouldExecute()
         return finder.HasPotentialNullableValueAccess;
     }
 
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var operationInstance = context.Operation.Instance;
+        if (operationInstance.Kind == OperationKindEx.PropertyReference
+            && operationInstance.ToPropertyReference() is { Instance: var instance } propertyReference
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:45]: The `var` decomposition is bit abusing the pattern matching when it doesn't have a null check. It's easier to track `property.Instance` 3 times.
`propertyReference` isn't wrong, this is just shorter and still good enough
```suggestion
            && operationInstance.ToPropertyReference() is { } property
```

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 15]
```diff
@@ -36,6 +38,33 @@ public override bool ShouldExecute()
...
+            && propertyReference.Property.Name == nameof(Nullable<int>.Value)
+            && IsNull(instance))
+        {
+            ReportIssue(instance, instance.Syntax.ToString());
+        }
+        else if (operationInstance.Kind == OperationKindEx.Conversion
+            && operationInstance.ToConversion() is { Operand: var operand } conversion
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:52]: Same as above
```suggestion
            && operationInstance.ToConversion() is { } conversion
```

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 16]
```diff
@@ -36,6 +38,33 @@ public override bool ShouldExecute()
...
+            && IsNull(operand))
+        {
+            ReportIssue(operand, operand.Syntax.ToString());
+        }
+
+        return context.State;
+
+        bool IsNull(IOperation operation) =>
+            context.HasConstraint(operation, ObjectConstraint.Null)
+            && SemanticModel.GetTypeInfo(operation.Syntax).Nullability().FlowState != NullableFlowState.NotNull;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:65]: What is this `FlowState` for? When I delete the line, no UT breaks. 

We either need to remove it, or add UT that will prove it has a value we can't detect.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
This looks pretty good

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 17]
```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
+            else
             {
-                Console.WriteLine(foo.Value.ToString());    // Compliant, unreachable
+                if (foo.HasValue)
+                {
+                    Console.WriteLine(foo.Value.ToString());    // Compliant, unreachable
+                }
+                if (!foo.HasValue)
+                {
+                    Console.WriteLine(foo.Value.ToString());    // FN, empty
```
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:405]: This was a bug in `ShouldExecute`, explained [here](https://github.com/SonarSource/sonar-dotnet/pull/6918#discussion_r1135962543) fixed in [step 5](https://github.com/SonarSource/sonar-dotnet/pull/6918).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 10 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:151]: Moved 19 spaces backward. 
Moving them 20 spaces would make the `;` symbol the farthest on the right and would leave no space between the semi-colon and the `//` characters (i.e. `.Value;// FN`), which in my opinion is not nice to see.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 18]
```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
-        return i.Value; //FIXME Non-compliant
+        _ = nullable.Value;                                              // Compliant, unknown
+        _ = (AStruct)nullable;                                           // Compliant, unknown
+        _ = new AStruct?(nullable.Value).Value;
+
+        nullable = null;
+        _ = nullable.Value;                                              // Noncompliant
+        nullable = new AStruct();
+        _ = nullable.Value;
+        nullable = new AStruct?(new AStruct());
```
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:159]: `new AStruct?(null);` doesn't compile: I can't find a constructor of `Nullable<T>` which takes a `T?` as parameter.

A test for `AStruct?()` has already been defined [here](https://github.com/SonarSource/sonar-dotnet/pull/6914/files#diff-9e6c038bf5ad1e53d1b547b8870231cc2887c9c53296f73e3eb811f51613bdedR162).

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 13 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:247]: The idea here was to test that the mere presence of a second nullable value doesn't affect the analysis of the first nullable value. The class is called `ComplexConditionMultipleNullables` and methods are called `DependentConditions`, where the most obvious way in which two variable can depend on each other is them being independent.

To make all this more explicit, I have:
- moved `DependentConditions2` to `ComplexConditionsSingleNullable` and changed into `bool LogicalAndLearningNonNull6(bool? b) => b.Value && d.HasValue;`. As you pointed out, having `b.Value` first is just another way of learning that `b` has a value.
- made more explicit `DependentConditions1` comment: `// Noncompliant, f presence doesn't affect d`.
- renumbered following dependent condition cases.

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 14 above)
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:45]: Ok for `property` instead of `propertyReference`.

I don't see why using pattern matching to deconstruct an object should be considered abuse.

In `switch` expressions, for instance, is quite natural and very readable:
```cs
string y = x switch
{
	(null, null) => "case 1",
	(var x1, var x2) when x1 == 4 && x2 == x1 + 1 => "case 2",
	_ => "case 3",
};
```

That is equivalent to 
```cs
string y;
if (x is (null, null)) y = "case 1";
else if (x is (var x1, var x2) && x1 == 4 && x2 == x1 + 1) y = "case 2";
else y = "case 3";
```

The most obvious way to match is to always match, which is what `var` does.

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 15 above)
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs:52]: I use `operand` to access its type, and it's not obvious to me that it can't be `null`.

So I haven't changed it to `operationInstance.ToConversion() is { Operand: { } operand } conversion`.

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 16 above)
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: I will reintroduce when and if I find a valid scenario for it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 10 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: "like -20" means roughly :)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs [hunk 19]
```diff
@@ -168,133 +144,275 @@ public static void LoopWithBreak(System.Collections.Generic.IEnumerable<string>
...
-public interface IWithDefaultImplementation
+class NullableOfCustomTypes
 {
-    int DoSomething()
+    void Assignment(AStruct? nullable)
     {
-        int? i = null;
-        return i.Value; //FIXME Non-compliant
+        _ = nullable.Value;                                              // Compliant, unknown
+        _ = (AStruct)nullable;                                           // Compliant, unknown
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: I don't think it tests the thing that was originally intended to be tested.

I think this was supposed to test cast when unknown value from argument is used. So we need to extract it into a separate method. Or re-assign it from some kind of `Unknown();` invocation.

Same for the line below with the constructor.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 11 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: More methods just adds more scaffolding in this case.

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 14 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: In switch expression, there's no other way. 
In the case we had, I have to track one additional state `instance` that represents just indirection/rename for something that we already have like `property.Instance`. That makes the code harder to read and understand.

The `var` should be used only where there's no other way.

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 15 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: Operations don't have `null` in their basic properties, otherwise they wouldn't be created.

Unless you can make it break with UT (or IT, or Peach validation later), there's no need for null check.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
No new comments, only unresolved above.

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 14 above)
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: So, if I understand it well, the indirection/rename is fine when it imposes some constraints on the capture (`case1` below), but it is not fine when it captures `instance` and only constraints the result of `ToPropertyReference` to be non-null, without constraining `Instance` to be non-null (`case2` below)?

```cs
case1: operationInstance.ToPropertyReference() is { Instance: { } instance }
case2: operationInstance.ToPropertyReference() is { Instance: var instance }
```

I fail to understand the logic behind such differentiation. Both pattern matching impose some constraints, with `case2` capturing a constrained property, whereas `case1` captures an unconstrained property (which is the goal of the var pattern feature in the first place).

The same indirection/rename also happens every time you create a local variable to store the result of a computation or casting to an object. The only difference is that the capture happens in an assignment statement.

```cs
var operationInstance = context.Operation.Instance;
// or
var switchExpression = (SwitchExpressionSyntaxWrapper)node;
var parent = (UnaryPatternSyntaxWrapper)node;
// etc.
```

Those cases are endemic in the `sonar-dotnet` code base.

I would argue that, in the `switch` expressions there is "no other way" just because pattern matching has been design to play together with deconstruction (an example of the design intents are explained by Neil Gafter in this post: https://github.com/dotnet/csharplang/issues/981#issuecomment-335952271), and that there is nothing wrong in `var` patterns.

And to conclude, I disagree with the statement that makes the code harder to read and understand.
It's easier for me to have a semantically well-defined variable for the rest of my code, rather than having property accesses like `property.Instance` scattered around. 
I think that is anyway a matter of personal taste, which should go beyond what should be presented as a required change in a pull request.

And on top, as mentioned in my previous comment, `.Instance` here is performing some computation, i.e. it is building a new wrapper every time you call it...

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: Changed, to use a local static method, either returning `null` or `new AStruct()`.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 11 above)
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: Removed the additional methods

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 14 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: There's a big difference between case 1 and 2. The `{ } instance` introduces a new state that is guaranteed to be not null and that changes the conditions under the condition is true. 
Contrary to `var instance` that doesn't have any functional effect and changes only the readability. I need to read a lot of syntax sugar to get something I already had and to learn that nothing happened. And I need to track one additional indirection. And that makes it harder to read.

For example `var` is fine here https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Rules/DoNotCallMethodsBase.cs#L42 where it's more practical than breaking it into 2 conditions and `var` statement.

Does it make sense?

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: I'm not concerned about the new instance, as shimming is generally a fixed price we have to pay to do this. And there will be roughly very few additional instances created (approximately as few as issues).

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

Please remove the `var` before merging.

File: analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs (hunk 14 above)
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs]: As discussed offline, I am changing both `ToPropertyReference` and `ToConversion` to store their results into dedicated variables (using `is var x`, which in this scenario is allowed) and calling `Instance` or `Operand` as many times as required.
